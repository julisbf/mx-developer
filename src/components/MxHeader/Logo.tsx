import React from 'react';
import {
    navigateByCallingMicroflow,
    SPRINTR,
    getCurrentApp,
    getEnvironmentLink,
} from '../../utils/mxHelpers';
import { connect } from '../../context/store';
import { links, microflows } from '../../resources/mendix.json';
import logoImage from '../../resources/img/mx_logo.png';

const currentApp = getCurrentApp();

interface LogoProps {
    loginUrl: string;
    loggedIn: boolean;
}

const Logo = ({ loginUrl, loggedIn }: LogoProps) => {
    const homeUrl = loggedIn ? getEnvironmentLink(links.home) : loginUrl;
    const navigateToHome = (event: React.MouseEvent<HTMLAnchorElement>) => {
        if (currentApp === SPRINTR) {
            event.preventDefault();
            navigateByCallingMicroflow(microflows.sprintr.home, homeUrl);
        }
    };
    return (
        <a href={homeUrl} onClick={navigateToHome} className="MxHeader__home">
            <img src={logoImage} alt="Logo" className="MxHeader__logo" />
        </a>
    );
};

export default connect()(React.memo(Logo));
