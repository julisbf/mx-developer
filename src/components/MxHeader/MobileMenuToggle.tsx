import React from 'react';

interface MobileMenuProps {
    toggle: (event: React.MouseEvent<HTMLButtonElement>) => void;
    isOn: boolean;
}

const MobileMenuToggle = ({ toggle, isOn }: MobileMenuProps) => (
    <button
        onClick={toggle}
        className="MxHeader__mobile-menu-toggle"
        type="button"
    >
        <span
            className={
                isOn
                    ? 'MxHeader__mobile-menu-toggle-box--on'
                    : 'MxHeader__mobile-menu-toggle-box'
            }
        >
            <span
                className={
                    isOn
                        ? 'MxHeader__mobile-menu-toggle-middle-bar--on'
                        : 'MxHeader__mobile-menu-toggle-middle-bar'
                }
            />
        </span>
    </button>
);

export default MobileMenuToggle;
