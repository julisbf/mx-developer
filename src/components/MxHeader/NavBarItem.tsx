import React from 'react';

import NavBarMenu, { Node } from './NavBarMenu';
import NavBarMenuToggle from './NavBarMenuToggle';
import { navigateByCallingMicroflow } from '../../utils/mxHelpers';
import { onSprintr } from '../../utils/environmentHelpers';

interface NavBarItemProps extends Node {
    isOnMobile: boolean;
    nodes?: Node[];
}

interface NavBarItemState {
    isMenuOpen: boolean;
}

/**
 * Use hyperlink or microflow
 * 1. if user is *on Sprintr*, use microflow to navigate *within* Sprintr
 *    because it's a lot faster than deeplink
 * 2. if the above condition is not met, always use deeplink (i.e. use hyperlink), e.g.
 *    - on Forum, go to Apps
 *    - on Sprintr, go to Apps
 *  */

class NavBarItem extends React.Component<NavBarItemProps, NavBarItemState> {
    state = { isMenuOpen: false };

    toggleMenu = (event: React.MouseEvent<HTMLElement>) => {
        event.stopPropagation();
        this.setState(({ isMenuOpen }) => ({ isMenuOpen: !isMenuOpen }));
    };

    render() {
        const {
            label,
            link,
            microflow,
            nodes,
            external,
            isOnMobile,
        } = this.props;
        const { isMenuOpen } = this.state;
        const navigate = () => navigateByCallingMicroflow(microflow, link);

        return (
            <div
                className={
                    isOnMobile
                        ? 'MxHeader__nav-bar-item'
                        : 'MxHeader__nav-bar-item--desktop'
                }
            >
                {onSprintr() || (isOnMobile && nodes && nodes.length > 0) ? (
                    <span
                        className="MxHeader__nav-bar-item-link"
                        onClick={navigate}
                        role="button"
                        onKeyPress={navigate}
                    >
                        {label}
                        {isOnMobile && (
                            <NavBarMenuToggle
                                isOpen={isMenuOpen}
                                onClick={this.toggleMenu}
                            />
                        )}
                    </span>
                ) : (
                    <a
                        href={link}
                        className={
                            isOnMobile
                                ? 'MxHeader__nav-bar-item-link'
                                : 'MxHeader__nav-bar-item-link--desktop'
                        }
                        target={external ? '_blank' : '_self'}
                    >
                        {label}
                    </a>
                )}
                {((isOnMobile && isMenuOpen) || !isOnMobile) && (
                    <NavBarMenu
                        nodes={nodes}
                        isOnMobile={isOnMobile}
                    ></NavBarMenu>
                )}
            </div>
        );
    }
}

export default NavBarItem;
