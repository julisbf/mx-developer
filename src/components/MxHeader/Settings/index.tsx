import React from 'react';

import { connect } from '../../../context/store';
import Search from './Search';
import Support from './Support';
import Notification from './Notification';
import Profile from './Profile';
import StartButton from './StartButton';
import LoginButton from './LoginButton';

const Settings = ({ loggedIn }: { loggedIn: boolean }) => (
    <div className="MxHeader__Setting">
        <Search />
        <Support />
        {loggedIn && <Notification />}
        {loggedIn && <Profile />}
        {!loggedIn && <StartButton />}
        {!loggedIn && <LoginButton />}
    </div>
);

export default connect()(React.memo(Settings));
