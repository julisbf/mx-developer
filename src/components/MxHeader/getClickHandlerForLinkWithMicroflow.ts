import React from 'react';

import { navigateByCallingMicroflow } from '../../utils/mxHelpers';

const getClickHandlerForLinkWithMicroflow = ({
    microflow,
    url,
    predicate,
    callback,
}: {
    microflow: string;
    url: string;
    predicate?: () => boolean;
    callback?: () => void;
}) => (event: React.MouseEvent<HTMLAnchorElement>) => {
    if ((predicate && predicate()) || !predicate) {
        event.preventDefault();
        navigateByCallingMicroflow(microflow, url);
        callback && callback();
    }
};

export default getClickHandlerForLinkWithMicroflow;
