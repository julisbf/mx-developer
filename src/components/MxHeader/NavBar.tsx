import React from 'react';

import { getCurrentApp, BEAVER, DATAHUB } from '../../utils/environmentHelpers';

import navbarItemData from '../../resources/menu/navbar.json';
import NavBarItem from './NavBarItem';

const NavBar = ({ isOpenOnMobile = false }: { isOpenOnMobile: boolean }) => {
    const currentApp = getCurrentApp();
    /**
     * Since we're using BEM, every css class is unique.
     * So, we should also make its implementation independent.
     * instead of requiring multiple classes `MxHeader__nav-bar MxHeader__nav-bar--mobile-menu-open`,
     * one class should be a self-contained state of the UI component, i.e. `MxHeader__nav-bar--mobile-menu-open`
     */
    const className = isOpenOnMobile
        ? 'MxHeader__nav-bar--mobile-menu-open'
        : 'MxHeader__nav-bar'; // use media-query to hide it on mobile when closed

    return (
        <div className={className}>
            {navbarItemData
                .filter(itemData => {
                    /**
                     * hide Sprintr-related NavBarItems on `beaver` (apps owned by digital ecosystem team)
                     *  */
                    return (
                        currentApp !== BEAVER ||
                        (currentApp === BEAVER && !itemData.microflow)
                    );
                })
                .filter(itemData => {
                    /**
                     * only show `Data Hub` when it's on `https://hub.mendixcloud.com`
                     *  */

                    return currentApp === DATAHUB || itemData.key !== DATAHUB;
                })
                .map(({ key, ...data }) => (
                    <NavBarItem
                        key={key}
                        {...data}
                        isOnMobile={isOpenOnMobile}
                    />
                ))}
        </div>
    );
};

export default NavBar;
