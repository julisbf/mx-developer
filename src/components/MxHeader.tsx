import React from 'react';

import Authenticate from './Authenticate';
import MobileMenuToggle from './MobileMenuToggle';
import Logo from './Logo';
import NavBar from './NavBar';
import Settings from './Settings';
import {
    getCurrentApp,
    BEAVER,
    SUPPORT,
    getWindowSize,
    PHONE,
    SCREEN_SM,
} from '../utils/mxHelpers';
import { Provider } from '../context/store';

interface MxHeaderProps {
    idTokenProviderMF?: string;
}

interface MxHeaderState {
    isMobileNavBarOpen: boolean;
}

class MxHeader extends React.Component<MxHeaderProps, MxHeaderState> {
    state = { isMobileNavBarOpen: false };

    componentDidMount() {
        window.addEventListener('resize', this.closeMobileMenuOnBigScreen);
    }

    componentWillUnmount() {
        window.removeEventListener('resize', this.closeMobileMenuOnBigScreen);
    }

    closeMobileMenuOnBigScreen = () => {
        const { isMobileNavBarOpen } = this.state;
        const windowSize = getWindowSize(window.innerWidth);
        const shouldShowMobileMenu = [PHONE, SCREEN_SM].includes(windowSize);
        if (!shouldShowMobileMenu && isMobileNavBarOpen)
            this.setState({ isMobileNavBarOpen: false });
    };

    toggleMobileNavBar = () =>
        this.setState(({ isMobileNavBarOpen }) => ({
            isMobileNavBarOpen: !isMobileNavBarOpen,
        }));

    render() {
        const { idTokenProviderMF } = this.props;
        const { isMobileNavBarOpen } = this.state;

        const currentApp = getCurrentApp();
        const showSettings = ![BEAVER, SUPPORT].includes(currentApp);

        const initialState = { idTokenProviderMF };
        return (
            <Provider initialState={initialState}>
                <div className="MxHeader__container">
                    <div className="MxHeader">
                        <Authenticate />
                        <MobileMenuToggle
                            toggle={this.toggleMobileNavBar}
                            isOn={isMobileNavBarOpen}
                        />
                        <Logo />
                        <NavBar isOpenOnMobile={isMobileNavBarOpen} />
                        <div className="MxHeader__white-space"></div>
                        {showSettings && <Settings />}
                    </div>
                </div>
            </Provider>
        );
    }
}

export default MxHeader;
