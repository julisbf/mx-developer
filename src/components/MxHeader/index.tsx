import React from 'react';
import debounce from 'tiny-debounce';

// import observe from '../../utils/observe';
import Authenticate from './Authenticate';
import MobileMenuToggle from './MobileMenuToggle';
import Logo from './Logo';
import NavBar from './NavBar';
import Settings from './Settings';
import { getCurrentApp, BEAVER, SUPPORT } from '../../utils/environmentHelpers';
import { getWindowSize, PHONE, SCREEN_SM } from '../../utils/screenSizeHelpers';
import { Provider } from '../../context/store';

interface MxHeaderProps {
    idTokenProviderMF?: string;
}

interface MxHeaderState {
    isMobileNavBarOpen: boolean;
}

class MxHeader extends React.Component<MxHeaderProps, MxHeaderState> {
    state = { isMobileNavBarOpen: false };

    componentDidMount() {
        window.addEventListener('resize', this.closeMobileMenuOnBigScreen);
    }

    componentWillUnmount() {
        window.removeEventListener('resize', this.closeMobileMenuOnBigScreen);
    }

    closeMobileMenuOnBigScreen = debounce(() => {
        const { isMobileNavBarOpen } = this.state;
        const windowSize = getWindowSize(window.innerWidth);
        const shouldShowMobileMenu = [PHONE, SCREEN_SM].includes(windowSize);
        if (!shouldShowMobileMenu && isMobileNavBarOpen)
            this.setState({ isMobileNavBarOpen: false });
    }, 100);

    toggleMobileNavBar = () =>
        this.setState(({ isMobileNavBarOpen }) => ({
            isMobileNavBarOpen: !isMobileNavBarOpen,
        }));

    render() {
        const { idTokenProviderMF } = this.props;
        const { isMobileNavBarOpen } = this.state;

        const currentApp = getCurrentApp();
        const showSettings = ![BEAVER, SUPPORT].includes(currentApp);

        const initialState = { idTokenProviderMF };
        return (
            <Provider initialState={initialState}>
                <div className="MxHeader__container">
                    <div className="MxHeader">
                        <Authenticate />
                        <MobileMenuToggle
                            toggle={this.toggleMobileNavBar}
                            isOn={isMobileNavBarOpen}
                        />
                        <Logo />
                        <NavBar isOpenOnMobile={isMobileNavBarOpen} />
                        <div className="MxHeader__white-space"></div>
                        {showSettings && <Settings />}
                    </div>
                </div>
            </Provider>
        );
    }
}

export default MxHeader;
