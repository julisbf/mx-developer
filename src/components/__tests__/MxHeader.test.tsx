import React from 'react';
import { render } from '@testing-library/react';

import MxHeader from '../MxHeader';

jest.mock('../../utils/getProfileViaAppBar2', () =>
    jest.fn(() => ({
        loginUrl: 'login-url',
        loggedIn: true,
    }))
);

jest.mock('../../utils/mxHelpers', () => ({
    getEnvironmentLink: jest.fn(link => link),
    getCurrentApp: jest.fn(() => 'forum'),
    callMicroflow: jest.fn(() => Promise.resolve('response-of-microflow')),
    onSprintr: jest.fn(() => true),
    onCloud: jest.fn(() => true),
    getAdminInfo: jest.fn(() =>
        Promise.resolve({
            hasCompanyAdmin: true,
            hasPlatformAdmin: true,
            hasOperationsDesk: true,
        })
    ),
}));

jest.mock('../../utils/parseIdToken', () =>
    jest.fn(() => ({
        username: 'mock-username',
        displayName: 'mock-displayName',
        email: 'mock-email',
        avatarThumbUrl: 'mock-avatarThumbUrl',
        avatarUrl: 'mock-avatarUrl',
        companyId: 'mock-companyId',
        companyName: 'mock-companyName',
        idToken: 'mock-idToken',
    }))
);

describe('MxHeader', () => {
    it('should render correctly', () => {
        const { container } = render(
            <MxHeader style={{}} idTokenProviderMF="idTokenProviderMF" />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
