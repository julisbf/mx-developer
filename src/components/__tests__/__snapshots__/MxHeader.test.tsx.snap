// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MxHeader should render correctly 1`] = `
<div
  class="MxHeader__container"
>
  <div
    class="MxHeader"
  >
    <button
      class="MxHeader__mobile-menu-toggle"
      type="button"
    >
      <span
        class="MxHeader__mobile-menu-toggle-box"
      >
        <span
          class="MxHeader__mobile-menu-toggle-middle-bar"
        />
      </span>
    </button>
    <a
      class="MxHeader__home"
    >
      <img
        alt="Logo"
        class="MxHeader__logo"
        src="mx_logo.png"
      />
    </a>
    <div
      class="MxHeader__nav-bar"
    >
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          Buzz
        </span>
      </div>
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          Apps
        </span>
      </div>
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          People
        </span>
      </div>
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          Community
        </span>
        <ul
          class="MxHeader__nav-bar-menu--desktop"
        >
          <li
            class="MxHeader__nav-bar-menu-item"
          >
            <a
              class="MxHeader__nav-bar-menu-item-link"
              href="https://developers.mendix.com/spotlight/"
              target="_self"
            >
              Blogs
            </a>
          </li>
          <li
            class="MxHeader__nav-bar-menu-item"
          >
            <a
              class="MxHeader__nav-bar-menu-item-link"
              href="https://developers.mendix.com/jobs/"
              target="_self"
            >
              Jobs
            </a>
          </li>
          <li
            class="MxHeader__nav-bar-menu-item"
          >
            <a
              class="MxHeader__nav-bar-menu-item-link"
              href="https://modelshare.mendix.com/"
              target="_self"
            >
              Model Share
            </a>
          </li>
          <li
            class="MxHeader__nav-bar-menu-item"
          >
            <a
              class="MxHeader__nav-bar-menu-item-link"
              href="https://developer.mendixcloud.com/link/mvp"
              target="_self"
            >
              MVP Program
            </a>
          </li>
          <li
            class="MxHeader__nav-bar-menu-item"
          >
            <a
              class="MxHeader__nav-bar-menu-item-link"
              href="https://developer.mendixcloud.com/link/leaderboards"
              target="_self"
            >
              Leaderboards
            </a>
          </li>
          <li
            class="MxHeader__nav-bar-menu-item"
          >
            <a
              class="MxHeader__nav-bar-menu-item-link"
              href="https://developer.mendixcloud.com/link/partneroverview"
              target="_self"
            >
              Our Partners
            </a>
          </li>
          <li
            class="MxHeader__nav-bar-menu-item"
          >
            <a
              class="MxHeader__nav-bar-menu-item-link"
              href="https://shop.mendix.com"
              target="_self"
            >
              Mendix Shop
            </a>
          </li>
        </ul>
      </div>
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          App Store
        </span>
      </div>
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          Academy
        </span>
      </div>
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          Forum
        </span>
      </div>
      <div
        class="MxHeader__nav-bar-item--desktop"
      >
        <span
          class="MxHeader__nav-bar-item-link"
          role="button"
        >
          Docs
        </span>
      </div>
    </div>
    <div
      class="MxHeader__white-space"
    />
    <div
      class="MxHeader__Setting"
    >
      <button
        class="MxHeader__search"
        id="mendix-header-search-button"
      >
        Search
      </button>
      <div
        class="MxHeader__support"
      >
        <ul
          class="MxHeader__support-menu"
        >
          <li
            class="MxHeader__support-menu-item"
          >
            <a
              class="MxHeader__support-menu-item-link"
              href="https://forum.mendixcloud.com/link/questions"
            >
              Ask the Community
            </a>
          </li>
          <li
            class="MxHeader__support-menu-item"
          >
            <a
              class="MxHeader__support-menu-item-link"
              href="https://docs.mendix.com"
            >
              Check the Documentation
            </a>
          </li>
          <li
            class="MxHeader__support-menu-item"
          >
            <a
              class="MxHeader__support-menu-item-link"
              href="https://support.mendix.com/hc/en-us"
            >
              Contact Mendix Support
            </a>
          </li>
        </ul>
      </div>
      <a
        class="MxHeader__start-button"
        href="https://signup.mendix.com/link/signup/?source=developers&medium=community&campaign=signup"
      >
        Start For Free
      </a>
      <a
        class="MxHeader__profile"
        title="Click here to login"
      >
        <img
          alt="Click here to login"
          class="MxHeader__avatar"
          src="avatar.png"
        />
      </a>
    </div>
  </div>
</div>
`;
